### ISSUE TITLE
Achieved close performance of MagicoderS by finetuning only with `evol-codealpaca-v1`.

### ISSUE THREAD

**Issue by wyt2000** [ise-uiuc/magicoder#31]
Thanks to your amazing tutorial, we reproduced the training process and experiments in the paper. The model finetuned by ourselves achieved the performance close to yours. For HumanEval(+), we got 57.32% / 52.44% pass@1 for Magicoder and 70.12% / 67.07% for MagicoderS.
Moreover, we conducted ablation studies to clarify the contribution of OSS-instruct with evol-instruct in the training process of MagicoderS. 

- We got close performance of MagicoderS (70.12% / 65.24% for HumanEval(+), similar results for DS-1000) by finetuning ONLY with `evol-codealpaca-v1` dataset under the same training setting mentioned in the tutorial. 
- We got even worse results (66.46% / 62.20% for HumanEval(+)) by swaping the training order of `oss-instruct-75k` and `evol-codealpaca-v1`.

We noticed that `oss-instruct-75k` was generated by base model `gpt-3.5-turbo-1106`, whereas `evol-codealpaca-v1` was generated by `gpt-4`, so the experiments of MagicoderS may be unfair. I think there should be additional evidence to show the contribution of OSS-instruct when it is combined with other data generation methods.

**Comment by UniverseFly** [comment]
Hi @wyt2000, thanks for reporting this!! Let me try to answer some of your questions.

We train on `oss-instruct (gpt-3.5-turbo-1106)` first and then `evol-codealpaca (gpt-4)` because `gpt-4` is much stronger than `gpt-3.5-turbo-1106`, and this progressive training can potentially help the model to learn better coding. Your interesting results demonstrated that this combination can still lead to an increase in HumanEval+, meaning the model generates more robust code. Though, I fully agree that there is plenty of room for further studies.

It's not surprising that reversing the order will harm the result because it is reversing the progressive learning process.

We hypothesize that the model to perform data generation can be very important, so we are also trying to generate `oss-instruct (gpt-4)`, but refining our approach to reduce the number of tokens `gpt-4` will generate as it is expensive. We hope to come out with more rigorous experiments to address your concerns.

Thank you again for these valuable experiments, and feel free to reach out to me through emails if you want to discuss more:)

**Comment by shangqing-liu** [comment]
Hello, everyone.

The discussion is valuable. I have tried to reproduce the training process with the standard shell scripts provided in README-DEV.md. I got similar results with @wyt2000. For HumanEval(+), I got 57.3 / 53.7 pass@1 but these values are lower than the paper reported (i.e., 60.4/55.5). May I seek for some help for further improvement? @UniverseFly 


Thank you.

**Comment by UniverseFly** [comment]
Hi @shangqing-liu , thanks for your question. You can reference https://github.com/ise-uiuc/magicoder/issues/28 to see the exact, recorded, training loss for each step. Hardware and hyperparameters can cause some randomness. So back then I just used 2xA100 and didn't set up flash attention.

**Comment by shangqing-liu** [comment]
Hi, thanks for the quick reply @UniverseFly and congrats on the ICML acceptance. When I turn off flash attention, I can get much closer results to the report value of HumanEval dataset in the arxiv version.

However, I have another doubt about using the provided scripts for MBPP dataset evaluation. I can get 0.667/0.563  values higher than the reported 0.642/0.526 when using Magicoder-CL-7B. So may I look for any explanation?


Thank you~

**Comment by UniverseFly** [comment]
Thanks @shangqing-liu. Recently, there has been an update of the EvalPlus framework ([this commit](https://github.com/evalplus/evalplus/commit/3de226896daa67b54e2683eb6e5f6840ce66018a) migrating from MBPP v0.1 to v0.2. Some problems are removed and bugs fixed. If you’re using v0.2, it’s likely to get higher results.

**Comment by shangqing-liu** [comment]
Thanks for the reply @UniverseFly . Can you tell us which version of eval plus you use? 0.1.?

Thank you.

**Comment by UniverseFly** [comment]
Yeah it’s v0.1 definitely (no v0.2 that time), but I might have not tracked the exact commit hash; there has been some bug fixes in between v0.1 commits.
